drivers/serial/uart_nrfx_uarte.c

```diff
@@ -0,0 +1,606 @@
+ *
+ * @param dev UARTE device struct
+ * @param baudrate Baud rate
+ *
+ * @return 0 on success or error code
+ */
+static int baudrate_set(struct device *dev, u32_t baudrate)
+{
+	nrf_uarte_baudrate_t nrf_baudrate; /* calculated baudrate divisor */
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	switch (baudrate) {
+	case 300:
+		/* value not supported by Nordic HAL */
+		nrf_baudrate = 0x00014000;
+		break;
+	case 600:
+		/* value not supported by Nordic HAL */
+		nrf_baudrate = 0x00027000;
+		break;
+	case 1200:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_1200;
+		break;
+	case 2400:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_2400;
+		break;
+	case 4800:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_4800;
+		break;
+	case 9600:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_9600;
+		break;
+	case 14400:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_14400;
+		break;
+	case 19200:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_19200;
+		break;
+	case 28800:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_28800;
+		break;
+	case 38400:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_38400;
+		break;
+	case 57600:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_57600;
+		break;
+	case 76800:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_76800;
+		break;
+	case 115200:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_115200;
+		break;
+	case 230400:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_230400;
+		break;
+	case 250000:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_250000;
+		break;
+	case 460800:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_460800;
+		break;
+	case 921600:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_921600;
+		break;
+	case 1000000:
+		nrf_baudrate = NRF_UARTE_BAUDRATE_1000000;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	nrf_uarte_baudrate_set(uarte, nrf_baudrate);
+
+	return 0;
+}
+
+/**
+ * @brief Poll the device for input.
+ *
+ * @param dev UARTE device struct
+ * @param c Pointer to character
+ *
+ * @return 0 if a character arrived, -1 if the input buffer is empty.
+ */
+static int uarte_nrfx_poll_in(struct device *dev, unsigned char *c)
+{
+	const struct uarte_nrfx_data *data = get_dev_data(dev);
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	if (!nrf_uarte_event_check(uarte, NRF_UARTE_EVENT_ENDRX)) {
+		return -1;
+	}
+
+	/* got a character */
+	*c = data->rx_data;
+
+	/* clear the interrupt */
+	nrf_uarte_event_clear(uarte, NRF_UARTE_EVENT_ENDRX);
+	nrf_uarte_task_trigger(uarte, NRF_UARTE_TASK_STARTRX);
+
+	return 0;
+}
+
+/**
+ * @brief Output a character in polled mode.
+ *
+ * @param dev UARTE device struct
+ * @param c Character to send
+ *
+ * @return Sent character
+ */
+static unsigned char uarte_nrfx_poll_out(struct device *dev,
+					 unsigned char c)
+{
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+
+	/* The UART API dictates that poll_out should wait for the transmitter
+	 * to be empty before sending a character. However, without locking,
+	 * this introduces a rare yet possible race condition if the thread is
+	 * preempted between sending the byte and checking for completion.
+
```

Comment:
This comment is misleading. The API dictates that the driver should send a character and wait until its transmitter is empty - true. There would be no race condition if there was a way to detect on the nRF52 that the transmitter indeed was empty so this alone does not lead to a race condition. 

Instead, the UART only supports events and an event flag is set when the transmitter has become empty. The race condition would be when higher-priority task cleared the flag at the end. Then the lower-priority task would never see this flag set and would stall. 

So perhaps this comment could be re-written to provide clarity?

Reply:
The API does not dictate that the driver should send a character and wait until its transmitter is empty. In fact it is oposite, first it shall check if transmitter is empty and next write a character to data register.
I have updated description with the information that UARTE supports only flag that transmitter become empty.

Reply:
What about rephrasing the first part to:
```
    /* The UART API dictates that poll_out should wait for the transmitter
     * to be empty before sending a character. However, the only way of
     * telling if the transmitter became empty in the UARTE peripheral is
     * to check if the ENDTX event for the previous transmission was set.
     * Since this event is not cleared automatically when a new transmission
     * is started, it must be cleared in software, and this leads to a rare
     * yet possible race condition if the thread is preempted right after
     * clearing the event but before sending a new character. The preempting
     * thread, if it also called poll_out, would then wait for the ENDTX
     * event that had no chance to become set.
```
Would this explain the issue better?